template_feed/Microsoft.DotNet.Common.ProjectTemplates.2.2/content/ConsoleApplication-FSharp/Program.fs

```diff
@@ -2,7 +2,13 @@
 
 open System
 
+// Define a function to print a message
```

Comment:
While i think is a good idea, to show declaration of functions BEFORE the main, i i think is that is annoying because that's template is also used **A LOT** for creating new applications, not just from new users, but for everyone.

That mean, we';ll need to remove that code **everytime**.
It's already annoying the `0  // return an integer exit code` at the end. but with also that will become really fast a noise

I think can be possibile to (weighted by my personal preference :D):

- or make the compiler error REALLY REALLY more helpful, if people declare the function AFTER usage. Atm is just `The value or constructor 'add' is not defined.`  but adding `if 'add' is a function, it must be defined before the usage, so before `main`
- or just add a comment to declare the functions BEFORE the main function, instead of declare a variable.
- or add a template option to make the main empty like `dotnet new console --empty`
- or add a template option (defaulted to behaviour for newbie users), to remove that. so like having and env var `FSHARP_TEMPLATE_BEGINNER=0`

Reply:
@cartermp added https://github.com/Microsoft/visualfsharp/issues/6677 to track error message, if can be improved

Reply:
The message improvement is good to have regardless, thanks for filing.

I'm much more keen on being "chatty" here in the templates than not. Yes, experienced F# devs would need to remove more stuff, but typically they're already removing nearly everything already. For example, the new library template also has a module and function that people typically remove.

I think the cost to pay is worth it so that people who are turned away by this don't just give up. We don't know exactly how many people try F# and fail, but after observing 9/9 people fail to grasp this concept with templates as-is, I felt something needed to be done.